Ar/folderstruct

AlexisRenchon opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
<!--- One sentence to describe the purpose of this PR, refer to any linked issues:
#14 -- this will link to issue 14
Closes #2 -- this will automatically close issue 2 on PR merge
-->
This is a small PR that we can hopefully merge soon (this week?) if you think it is a good idea (it is okay if you think it is not!)

The idea is to have a documentation page that describes the structure of `ClimaLSM` repository. When I go through a repo on GitHub, I always try to understand what each folder and each file is meant to do, it helps me navigate and understand how the repo is structured and work... I imagine it could be very helpful for newcomers, whether they are students or advanced scientist. It can also help us organize the repo better. 

This PR goes hand-to-hand with PR #220 but is small and easy to merge (it is just a doc page).  

## To-do
<!---  list of proposed tasks in the PR, move to "Content" on completion 
- Proposed task
-->
Just go through https://github.com/CliMA/ClimaLSM.jl/blob/ar/folderstruct/docs/src/folderstruct.md, it is easy to read it there, and edit it as you see fit. 

## Content
<!---  specific tasks that are currently complete 
- Solution implemented
-->
Currently it is very simplified. We can decide how much details we go into, let's discuss on this PR below! 

<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [x] I have read and checked the items on the review checklist.

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+
+This documentation page is meant to help new users navigate through the repository of [ClimaLSM](https://github.com/CliMA/ClimaLSM.jl).
+
+`ClimaLSM` home directory has 5 main folders:<br>
```
can remove the <br> (it appears for me in the rendered page?)

juliasloan25 replied:
I think you mean to remove the \<br\> (it got removed from the markdown comment). It doesn't look like it shows up in the rendered page for me though (looking at "Preview")

kmdeck commented:
hey Alexis! I think this is a nice idea, I'll leave a few comments in review.

We might just be careful to not describe too much in detail because then this doc page would need to get updated often, and it is easy to forget to. But right now it looks like a good level!

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+- experiments: contains simple runs of `ClimaLSM` models. 
+- parameters: contains a file to retrieve constants such as avogadro's number, the speed of light, etc. 
+- src: contains the code of `ClimaLSM` models. 
+- test: contains [unit tests](https://en.wikipedia.org/wiki/Unit_testing), which are meant to ensure small pieces of ClimaLSM source code works as intended before merging pull requests.
```
"work as intended"

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+- test: contains [unit tests](https://en.wikipedia.org/wiki/Unit_testing), which are meant to ensure small pieces of ClimaLSM source code works as intended before merging pull requests.
+
+and 3 GitHub actions folders. GitHub actions are `.yml` files, which are bash scripts that runs on a remote computer on each git push. 
+- .builkite: contains a script building outputs such as figures from experiments and tests folders. 
```
buildkite (small typo)

We can also note that these buildkite runs are carried out as part of CI and must run without error in order to merge a PR.

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+
+and 3 GitHub actions folders. GitHub actions are `.yml` files, which are bash scripts that runs on a remote computer on each git push. 
+- .builkite: contains a script building outputs such as figures from experiments and tests folders. 
+- .dev: contains
```
i think right now this just has the julia formatter? Not totally sure how to describe the purpose of this folder.

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+The `experiments` folder contains scripts to run ClimaLSM models. It contains a folder for `integrated` models and a folder for `standalone` models. It is meant to provide users with simple examples of ClimaLSM runs. The files contains meteorological inputs (such as precipitation), values for every parameters, and the domains and timestepper are specified. 
+
+For example, `/experiments/LSM/ozark/` contains:
+- ozark_domains.jl: Describes the soil domain (depth, number of layer), and the canopy (number and height of stems and leaves).
```
ozark_domain.jl

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+For example, `/experiments/LSM/ozark/` contains:
+- ozark_domains.jl: Describes the soil domain (depth, number of layer), and the canopy (number and height of stems and leaves).
+- ozark_met_drivers_FLUXNET.jl: This files load meteorological input data from the ozark FLUXNET file, and does additional things such as spline interpolation of these drivers.
+- ozark_parameters.jl: In this file, parameters values are defined, for example the soil porosity is given a value of 0.55 m3 m-3. 
```
here we may not want to give the specific value, because then any PR which changes that value also has to remember to update this docs page

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+## /src folder
+
+The `/src` folder contains the source code of `ClimaLSM` models. It contains 3 folders:
+- SharedUtilities: This is a core folder that defines functions to create `ClimaLSM` models, and many utility functions to be called across the repository. For example, `SharedUtilities/models.jl` defines and export the function `make_update_aux` which will be used to store auxiliary variables in a nested variable `p`, `SharedUtilities/boundary_conditions.jl` defines functions for setting boundar ...
```
for PDE domains

kmdeck replied:
"which will be used to create a function which updates the auxiliary parameters, stored in the vector `p`"

kmdeck commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+
+As well as 3 files:
+- ClimaLSM.jl: This file is the main Julia module of `ClimaLSM.jl` repository. It contains all functions defined in `/src` in a nested way, for example `ClimaLSM.X`, `ClimaLSM.Soil.X`, 'ClimaLSM.Canopy.X`, etc. When a Julia user install and uses ClimaLSM via `]add ClimaLSM, using ClimaLSM`, they are loading those functions, and are ready to use ClimaLSM codebase. 
+- Parameters.jl: 
```
This is how we interact with CLIMAParameters, which describes a single interface and parameter set that all Clima models use. This ensures all components of the Earth System Model use the same values for all parameters, and that the same parameter calibration tools can be used to calibrate all models, either jointly or independently.

juliasloan25 commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+As well as 3 files:
+- ClimaLSM.jl: This file is the main Julia module of `ClimaLSM.jl` repository. It contains all functions defined in `/src` in a nested way, for example `ClimaLSM.X`, `ClimaLSM.Soil.X`, 'ClimaLSM.Canopy.X`, etc. When a Julia user install and uses ClimaLSM via `]add ClimaLSM, using ClimaLSM`, they are loading those functions, and are ready to use ClimaLSM codebase. 
+- Parameters.jl: 
+- Regridder.jl:
```
Contains code used to regrid fields between different meshes, using ClimaCore data structures. This code uses ClimaCoreTempestRemap.jl under the hood and is largely copied from ClimaCoupler.jl. It may be moved to a different CliMA repository in the near future.

This code could also probably be moved to SharedUtilities

AlexisRenchon replied:
Same comment here as for Parameters.jl. Move to SharedUtilities @kmdeck?

juliasloan25 commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+- Standalone: This folder contains standalone models, which are submodels that can be run independently of each other. This is an important aspect of `ClimaLSM` code design: to maximise modularity, sub-models can be run alone, and many different methods of the same sub-model can be defined via Julia multiple-disptach. 
+- Integrated: This folder contains integrated models. It assembles standalone models together, as one would assemble pieces of a puzzle. Thanks to the modularity of `ClimaLSM` design, many configuration of LSM can be assembled in integrated models. 
+
+As well as 3 files:
```
There are also 3 other files - `pond_soil_model.jl`, `soil_energy_hydrology_biogeochemistry.jl`, and `soil_plant_hydrology_model.jl`. Are they not included here because we're planning to move them to Standalone/Integrated folders?

AlexisRenchon replied:
Yes they will be in integrated folder. 
I wonder if it is okay to merge this PR first though, even if it will have some inconsistencies until we actually merge #220 (that needs some work before being merged)

kmdeck replied:
I think it is ok! the intention is to follow up with the restructure relatively quickly, so all good.

juliasloan25 reviewed (commented):
This is a great idea! Thanks for doing this, Alexis. I left a few comments/suggestions. Another thought is maybe we could rename the file itself to `folderstructure.md` - I think that would be a bit more clear

AlexisRenchon commented on docs/src/folderstruct.md:
```diff
@@ -0,0 +1,55 @@
+
+As well as 3 files:
+- ClimaLSM.jl: This file is the main Julia module of `ClimaLSM.jl` repository. It contains all functions defined in `/src` in a nested way, for example `ClimaLSM.X`, `ClimaLSM.Soil.X`, 'ClimaLSM.Canopy.X`, etc. When a Julia user install and uses ClimaLSM via `]add ClimaLSM, using ClimaLSM`, they are loading those functions, and are ready to use ClimaLSM codebase. 
+- Parameters.jl: 
```
I will leave this comment unresolved, as maybe we should indeed move Parameters.jl into SharedUtilities. (We could do it in this PR?) 
What do you think @kmdeck?

kmdeck replied:
I think it is a good idea!

kmdeck replied:
we can do it in this PR, I think!

AlexisRenchon replied:
ok! will do in next commit =)

AlexisRenchon commented:
Hello @juliasloan25, @kmdeck, I have been through your comments and edited the PR accordingly! 

I see on the docs preview that there are still some small issues to fix (e.g., some text in italic not intentionally).

Julia mentioned moving `/src/Parameters.jl` and `/src/Regridder.jl` to `src/SharedUtilities/`, which would be neat from an organization perspective (less files in /src - /src would just be SharedUtilities, Standalone, Integrated folder and ClimaLSM.jl!).   

Moving files is easy, but that means we need to adjust the path to these files in the scripts calling them. I can do it in the next commit if you agree it is a good idea @kmdeck. 

After that, maybe someone else could do a final review before we merge? 

Then I will also finalize #220 - hopefully all done early next week!

AlexisRenchon commented on docs/src/folderstructure.md:
```diff
@@ -0,0 +1,56 @@
+
+The `/src` folder contains the source code of `ClimaLSM` models. It contains 3 folders:
+- SharedUtilities: This is a core folder that defines functions and data structures used across all modules and models types of `ClimaLSM`. For example, `SharedUtilities/models.jl` defines and export the function `make_update_aux` which will be used to create a function which updates the auxiliary parameters, stored in the vector `p`, `SharedUtilities/boundary_conditions.jl` defines functions for ...
+- Standalone: This folder contains standalone models, which are submodels that can be run independently of each other. This is an important aspect of `ClimaLSM` code design: to maximise modularity, sub-models can be run alone, and many different methods of the same sub-model can be defined via Julia multiple-disptach. 
```
I think it is a good thing to mention!

AlexisRenchon commented:
I think we are ready to merge!

I moved `src/Regridder.jl` and `src/Parameters.jl` to `src/SharedUtilities/Regridder.jl` and `src/SharedUtilities/Parameters.jl`. I corrected the path to these files in `src/ClimaLSM.jl`. I believe it should be all set. If it passes CI it should mean it is okay?

kmdeck reviewed (approved):
looks good to me! But first, can we squash these 17 commits into 1? I can help with that if you want to pair-program over zoom!

kmdeck commented on docs/src/folderstructure.md:
```diff
@@ -0,0 +1,54 @@
+
+As well as 3 files:
+- ClimaLSM.jl: This file is the main Julia module of `ClimaLSM.jl` repository. It contains all functions defined in `/src` in a nested way, for example `ClimaLSM.X`, `ClimaLSM.Soil.X`, 'ClimaLSM.Canopy.X`, etc. When a Julia user install and uses ClimaLSM via `]add ClimaLSM, using ClimaLSM`, they are loading those functions, and are ready to use ClimaLSM codebase. 
+- Parameters.jl: This script describes a single interface and parameter set that all CliMA models use. This ensures all components of the Earth System Model use the same values for all parameters, and that the same parameter calibration tools can be used to calibrate all models, either jointly or independently.
```
oh! one other quick fix - these descriptions can now be moved under "SharedUtilities", since you moved them :)

AlexisRenchon commented:
bors r+
